core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -615,17 +618,16 @@ private[log] class Cleaner(val id: Int,
     // but we may be able to fit more (if there is lots of duplication in the dirty section of the log)
     var offset = dirty.head.baseOffset
     require(offset == start, "Last clean offset is %d but segment base offset is %d for log %s.".format(start, offset, log.name))
+
     val maxDesiredMapSize = (map.slots * this.dupBufferLoadFactor).toInt
-    var full = false
-    for (segment <- dirty if !full) {
+    var didPut = true
+    for (segment <- dirty if (map.size < maxDesiredMapSize && didPut)) {
```

Comment:
Using the `if` inside the `for` means that `withFilter` is called on the collection with that predicate. It is best used when there are no side-effects, but here we are relying on the mutation happening in the loop itself. I suggest you use a plain `if` inside the loop.